From 63c4718affaefc0061ab28cbd22986fe30251b4a Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Fri, 26 Oct 2018 11:50:13 +0300
Subject: [PATCH 4937/5242] MLK-20098-1: imx/drm/dcss: Fix potential sleep in
 IRQ context

commit  f34773489929e1b0d79e1e56686bee048591f645 from
https://source.codeaurora.org/external/imx/linux-imx.git

Activating CONFIG_SLEEP_ATOMIC_SLEEP detected a couple of potential sleeps
inside IRQ context:

[   23.609203] BUG: sleeping function called from invalid context at kernel/irq/manage.c:112
[   23.617437] in_atomic(): 1, irqs_disabled(): 128, pid: 0, name: swapper/2
[   23.624229] CPU: 2 PID: 0 Comm: swapper/2 Tainted: G        W 4.14.62-05295-gf2fa7e6 #454
[   23.632927] Hardware name: Freescale i.MX8MQ EVK (DT)
[   23.637980] Call trace:
[   23.640433] [<ffff00000808a360>] dump_backtrace+0x0/0x3d8
[   23.645834] [<ffff00000808a74c>] show_stack+0x14/0x20
[   23.650891] [<ffff000008dba640>] dump_stack+0x9c/0xbc
[   23.655946] [<ffff0000080f70d4>] ___might_sleep+0xf4/0x118
[   23.661433] [<ffff0000080f7148>] __might_sleep+0x50/0x88
[   23.666750] [<ffff00000811f218>] synchronize_irq+0x30/0x98
[   23.672237] [<ffff00000811f6e8>] disable_irq+0x20/0x30
[   23.677378] [<ffff00000866edb8>] dcss_dpr_irq_enable+0x78/0x98
[   23.683211] [<ffff00000866f798>] dcss_dtg_vblank_irq_enable+0x40/0x78
[   23.689652] [<ffff00000866c79c>] dcss_vblank_irq_enable+0xc/0x18
[   23.695661] [<ffff0000086d3048>] dcss_disable_vblank+0x30/0x50
[   23.701496] [<ffff0000086aaa2c>] drm_vblank_disable_and_save+0xd4/0xe8
[   23.708023] [<ffff0000086aaac8>] vblank_disable_fn+0x88/0xa8
[   23.713685] [<ffff00000813513c>] call_timer_fn.isra.5+0x24/0x80
[   23.719603] [<ffff00000813523c>] expire_timers+0xa4/0xb0
[   23.724914] [<ffff000008135300>] run_timer_softirq+0xb8/0x170
[   23.730660] [<ffff000008081bcc>] __do_softirq+0x12c/0x228
[   23.736062] [<ffff0000080d57bc>] irq_exit+0xc4/0x100
[   23.741025] [<ffff00000811e528>] __handle_domain_irq+0x60/0xb8
[   23.746857] [<ffff000008081998>] gic_handle_irq+0x78/0x17c

These sleep warnings were generated because disable_irq() may sleep. Use
disable_irq_nosync() instead.

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dcss/dcss-dpr.c |    6 +++---
 drivers/gpu/imx/dcss/dcss-dtg.c |    2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/imx/dcss/dcss-dpr.c b/drivers/gpu/imx/dcss/dcss-dpr.c
index 03663533..cdaf582 100644
--- a/drivers/gpu/imx/dcss/dcss-dpr.c
+++ b/drivers/gpu/imx/dcss/dcss-dpr.c
@@ -736,9 +736,9 @@ void dcss_dpr_irq_enable(struct dcss_soc *dcss, bool en)
 	struct dcss_dpr_priv *dpr = dcss->dpr_priv;
 
 	if (!en) {
-		disable_irq(dpr->ch[0].irq);
-		disable_irq(dpr->ch[1].irq);
-		disable_irq(dpr->ch[2].irq);
+		disable_irq_nosync(dpr->ch[0].irq);
+		disable_irq_nosync(dpr->ch[1].irq);
+		disable_irq_nosync(dpr->ch[2].irq);
 
 		return;
 	}
diff --git a/drivers/gpu/imx/dcss/dcss-dtg.c b/drivers/gpu/imx/dcss/dcss-dtg.c
index 7598403..47f43cc 100644
--- a/drivers/gpu/imx/dcss/dcss-dtg.c
+++ b/drivers/gpu/imx/dcss/dcss-dtg.c
@@ -542,7 +542,7 @@ void dcss_dtg_ctxld_kick_irq_enable(struct dcss_soc *dcss, bool en)
 	if (!dtg->ctxld_kick_irq_en)
 		return;
 
-	disable_irq(dtg->ctxld_kick_irq);
+	disable_irq_nosync(dtg->ctxld_kick_irq);
 	dtg->ctxld_kick_irq_en = false;
 }
 EXPORT_SYMBOL(dcss_dtg_ctxld_kick_irq_enable);
-- 
1.7.9.5

